core/src/main/java/it/pagopa/selfcare/mscore/core/InstitutionServiceImpl.java

```diff
@@ -287,36 +287,62 @@ public void checkIfAlreadyExists(String externalId) {
         }
     }
 
-    private List<RelationshipInfo> toRelationshipInfo(List<OnboardedUser> institutionRelationships, Institution institution) {
+    private List<RelationshipInfo> toRelationshipInfo(List<OnboardedUser> institutionRelationships, Institution institution, List<PartyRole> roles, List<RelationshipState> states, List<String> products, List<String> productRoles) {
         List<RelationshipInfo> list = new ArrayList<>();
         for (OnboardedUser onboardedUser : institutionRelationships) {
             for (UserBinding binding : onboardedUser.getBindings()) {
-                retrieveAllProduct(list, onboardedUser.getId(), binding, institution);
+                list.addAll(retrieveAllProduct(onboardedUser.getId(), binding, institution, roles, states, products, productRoles));
             }
         }
         return list;
     }
 
-    protected void retrieveAllProduct(List<RelationshipInfo> list, String userId, UserBinding binding, Institution institution) {
+    protected List<RelationshipInfo> retrieveAllProduct(String userId, UserBinding binding, Institution institution, List<PartyRole> roles, List<RelationshipState> states, List<String> products, List<String> productRoles) {
+        List<RelationshipInfo> list = new ArrayList<>();
         if (institution != null) {
             if (institution.getId().equalsIgnoreCase(binding.getInstitutionId())) {
-                for (OnboardedProduct product : binding.getProducts()) {
+                list = binding.getProducts().stream()
+                        .filter(product -> filterProduct(product, roles, states, products, productRoles))
+                        .map(product -> {
+                            RelationshipInfo relationshipInfo = new RelationshipInfo();
+                            relationshipInfo.setInstitution(institution);
+                            relationshipInfo.setUserId(userId);
+                            relationshipInfo.setOnboardedProduct(product);
+                            return relationshipInfo;
+                        })
+                        .collect(Collectors.toList());
+            }
+        } else {
+            for (OnboardedProduct product : binding.getProducts()) {
+                if (filterProduct(product, roles, states, products, productRoles)) {
+                    Institution inst = retrieveInstitutionById(binding.getInstitutionId());
                     RelationshipInfo relationshipInfo = new RelationshipInfo();
-                    relationshipInfo.setInstitution(institution);
+                    relationshipInfo.setInstitution(inst);
```

Comment:
non usare abbreviazioni quando puoi dare nomi più parlanti, ricorda sempre che il codice deve essere ben comprensibile. 
Meglio aggiungere qualche carattere in più che in meno ;)